src/Avalonia.Native/IAvnMenu.cs

```diff
@@ -0,0 +1,176 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.Reactive.Disposables;
+using Avalonia.Controls;
+using Avalonia.Platform.Interop;
+
+namespace Avalonia.Native.Interop
+{
+    class MenuEvents : CallbackBase, IAvnMenuEvents
+    {
+        private IAvnMenu _parent;
+
+        public void Initialise(IAvnMenu parent)
+        {
+            _parent = parent;
+        }
+
+        public void NeedsUpdate()
+        {
+            _parent?.RaiseNeedsUpdate();
+        }
+    }
+
+    public partial class IAvnMenu
+    {
+        private MenuEvents _events;
+        private AvaloniaNativeMenuExporter _exporter;
+        private List<IAvnMenuItem> _menuItems = new List<IAvnMenuItem>();
+        private Dictionary<NativeMenuItemBase, IAvnMenuItem> _menuItemLookup = new Dictionary<NativeMenuItemBase, IAvnMenuItem>();
+        private CompositeDisposable _propertyDisposables = new CompositeDisposable();
+
+        internal void RaiseNeedsUpdate()
+        {
+            (ManagedMenu as INativeMenuExporterEventsImplBridge).RaiseNeedsUpdate();
+
+            _exporter.InvalidateMenu();
```

Comment:
We don't need to *enqueue* invalidation pass here. We need to run one if it was already enqueued by the user code as a reaction to `Opening` event.

Reply:
```
internal void InvalidateMenu()
        {
            if (_resetQueued)
            {
                DoLayoutReset();
            }
        }
```

Bad naming of method?

Reply:
Yes, invalidate = mark as invalid. It usually doesn't trigger any actions immediately. `UpdateIfNeeded` should be better.

Reply:
Renamed.